### ISSUE TITLE
Revealed statements representation

### ISSUE THREAD

**Issue by tplooker** [w3c-ccg/ldp-bbs2020#22]
Currently in a `BBSSignatureProof2020` we represent the index's of the messages that were revealed and the positions they occupied in the originally signed message set as an array, see [here](https://mattrglobal.github.io/jsonld-signatures-bbs-spec/#revealed-indicies) for more details. However a more compact representation of this that could be included in the actual `proof` value of the proof as an encoded bit string. Say for example 10 statements were signed and the resulting proof revealed the first five messages, the corresponding bit stream in binary would be `1111100000`, because this would be a variable in length based on the total statements signed we could also elect to preappend the totalStatements value in the proof encoded as a say a 16bit unsigned int giving us a maximum of 65535 total statements in a single proof.

This would for example change the following proof from 

```
{
  .....
  "proof": {
    "type": "BbsBlsSignatureProof2020",
    "created": "2020-04-26T04:21:07Z",
    "verificationMethod": "did:example:489398593#test",
    "proofPurpose": "assertionMethod",
    "proof": "GO/i24loDTTgUtMCGM/jivlD260k93d9ek2FxB/L2NQmZANjKd13r+8yDIrRqD5hB1HjIc1gY3Y/lwexZNUa+BAlaXBQZa8iXhYr6M9lMauZxlYuaJkyGlUiu/QwcHwDD9hpB2LK8kUvQHdYZmkBwL3Whyqptl2hkgNIdCnpqoBH+L9DmIZH9iGwrzYJ6rx/AAAAdIu4GCCrIhQ1Vb/BOlHYaer1eTC+Sukw3ypVmNNoP5pa4nDD+/UQin2HjmS8YtZNqwAAAAIMTkAU/oDTR4EcEsFcUbiM9ThFKytMZ/uGC28463I/9Bb1JAL3F23JgUHe5eJzScg7Nu2hDHpkskO4/NaExd0cA/Sle9qeoObCi6trWtR8+ZuAI5tpkiOojsKBGoU55Diy8rMyOA+ruP2+F6tptRavAAAACUiVcbDWpO7LE8hMFmAfrO+DrWd7S0T2opAk6QheOTdnUfZIO5gpCDEvXGnZ2pmnGYqcLnAjth/gwhAEfTSTrntnAJQomBJStgVS73wZgVTgchhOFO2Qbg0nFGHyhzkZJD0NSjihxQDOx45pYSaIqiF0uM4iGLh79G5xU2Av+PBqbG4ASU1kzXa8N2cE6F7osl5LYKvm+yeGl2gDktCRwrcansuuZ4RQpPFE2ZCNYlsPKrjBQ+QTgWjpHQmhtfLpVFJcFpIy4x2GUD3tkZFGKYpEm2Sc00bNzfYozLdKj4erTr17SjoHwYyHwiofPmb2PRcrknpYVJaxyrVYM9sn9gwEoI4dLJRbT66k4lgEXRJJ3Kb430WHyWuWABmJ2dd0iQ==",
    "revealStatements": [
      0,
      1,
      2,
      4,
      5,
      6,
      8,
      9,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23
    ],
    "totalStatements": 24,
    "nonce": "M/e44JTNSsfhnykE0yoD8eaYIdJARbpDIWFhu+TWwc70J5iwPHa8Q6bYQd1YjjxpV4c="
  }
}
```

To

```
{
  .....
  "proof": {
    "type": "BbsBlsSignatureProof2020",
    "created": "2020-04-26T04:21:07Z",
    "verificationMethod": "did:example:489398593#test",
    "proofPurpose": "assertionMethod",
    "proof": " {binary2base64('0000000000100100111011101100000111111111')}GO/i24loDTTgUtMCGM/jivlD260k93d9ek2FxB/L2NQmZANjKd13r+8yDIrRqD5hB1HjIc1gY3Y/lwexZNUa+BAlaXBQZa8iXhYr6M9lMauZxlYuaJkyGlUiu/QwcHwDD9hpB2LK8kUvQHdYZmkBwL3Whyqptl2hkgNIdCnpqoBH+L9DmIZH9iGwrzYJ6rx/AAAAdIu4GCCrIhQ1Vb/BOlHYaer1eTC+Sukw3ypVmNNoP5pa4nDD+/UQin2HjmS8YtZNqwAAAAIMTkAU/oDTR4EcEsFcUbiM9ThFKytMZ/uGC28463I/9Bb1JAL3F23JgUHe5eJzScg7Nu2hDHpkskO4/NaExd0cA/Sle9qeoObCi6trWtR8+ZuAI5tpkiOojsKBGoU55Diy8rMyOA+ruP2+F6tptRavAAAACUiVcbDWpO7LE8hMFmAfrO+DrWd7S0T2opAk6QheOTdnUfZIO5gpCDEvXGnZ2pmnGYqcLnAjth/gwhAEfTSTrntnAJQomBJStgVS73wZgVTgchhOFO2Qbg0nFGHyhzkZJD0NSjihxQDOx45pYSaIqiF0uM4iGLh79G5xU2Av+PBqbG4ASU1kzXa8N2cE6F7osl5LYKvm+yeGl2gDktCRwrcansuuZ4RQpPFE2ZCNYlsPKrjBQ+QTgWjpHQmhtfLpVFJcFpIy4x2GUD3tkZFGKYpEm2Sc00bNzfYozLdKj4erTr17SjoHwYyHwiofPmb2PRcrknpYVJaxyrVYM9sn9gwEoI4dLJRbT66k4lgEXRJJ3Kb430WHyWuWABmJ2dd0iQ==",
    "nonce": "M/e44JTNSsfhnykE0yoD8eaYIdJARbpDIWFhu+TWwc70J5iwPHa8Q6bYQd1YjjxpV4c="
  }
}
```

**Comment by kdenhartog** [comment]
The proof example you show here doesn't seem to align with what you're showing in the bytes as the example. `0000000000100100111011101100000111111111` is longer than 16 bytes and 24 bytes (total number of statements), so I'm not sure how this is being bunched together. In any case If we're going to prepend this on to the proof, we need to either indicate length or set a fixed length as well so we can properly parse the reveal statements from the actual proof.

Also, I'm not following how the reveal statements is indicated with this example. Could you provide further detail?

**Comment by tplooker** [comment]
So this example is showing a 16bit fix length representation of the totalStatements, followed by a bit string of length indicated by the first 16bits so it should be 16+24 = 40 bits

**Comment by tplooker** [comment]
E.g the first 16bit indicate there are 24 statements therefore the next 24 bits indicate the reveal, or not revealed statuses of the messages

**Comment by tplooker** [comment]
This analagous to how a message length prefix is pre-appended in the serializations of protocol buffers.

**Comment by OR13** [comment]
I like this suggestion, the more we can hide the details the better:)

**Comment by kdenhartog** [comment]
So to process this we'll need to parse off the first 16 bytes, identify the variable length following that and then parse off X trailing bytes? Seems easy enough to do and a worth improvement now that I understand it clearly.

**Comment by tplooker** [comment]
If by bits you mean bytes then absolutely :)

**Comment by mikelodder7** [comment]
I'm in favor. Never fear having more than 64K statements, the code could probably handle it but I don't think its realistic.

**Comment by tplooker** [comment]
The outstanding question here to address is that the ordering of the statements in the normalized form of the derived proof is preserved in reference to the order of the statements signed in the original proof. 

For instance, if some of the statements in the derived proof are re-ordered when normalized then the proof validation.

This is related to #10 and the guarantees we will have about the normalized form of the derived proof

**Comment by tplooker** [comment]
This implementation has been made in [node-bbs-signatures](https://github.com/mattrglobal/node-bbs-signatures) and will be available in the 0.8.0 release, the required changes to [jsonld-bbs-signatures](https://github.com/mattrglobal/jsonld-signatures-bbs) are in progress
